source/Detector/Amptek/AmptekSDD123Server.cpp

```diff
@@ -0,0 +1,570 @@
+		lastRequestSpectrumTime_->start();
+		replySpectrumTime_ = new QTime();
+		replySpectrumTime_->start();
+
+		udpDetectorSocket_ = new QUdpSocket(this);
+
+		connect(udpDetectorSocket_, SIGNAL(readyRead()), this, SLOT(processPendingDatagrams()));
+		connect(this, SIGNAL(responsePacketReady()), this, SLOT(onResponsePacketReady()));
+	}
+}
+
+bool AmptekSDD123Server::detectorUnavailable() const{
+	return detectorUnavailable_;
+}
+
+QList<double> AmptekSDD123Server::droppedPacketStatistics() const{
+	QDateTime last30Seconds = QDateTime::currentDateTime().addSecs(-30);
+	QList<double> retVal;
+	retVal << droppedPackets_.totalOccurances() << droppedPackets_.totalRate() << droppedPackets_.occurancesSince(last30Seconds) << droppedPackets_.rateSince(last30Seconds);
+	return retVal;
+}
+
+QList<double> AmptekSDD123Server::unrequestedPacketStatistics() const{
+	QDateTime last30Seconds = QDateTime::currentDateTime().addSecs(-30);
+	QList<double> retVal;
+	retVal << unrequestedPackets_.totalOccurances() << unrequestedPackets_.totalRate()<< unrequestedPackets_.occurancesSince(last30Seconds) << unrequestedPackets_.rateSince(last30Seconds);
+	return retVal;
+}
+
+bool AmptekSDD123Server::event(QEvent *e){
+	if(e->type() == (QEvent::Type)AmptekEventDefinitions::ConfigurationInitiateRequestEvent){
+		if(((AmptekConfigurationInitiateRequestEvent*)e)->configurationMode_){
+			emit serverAboutToChangeToConfigurationState();
+			QTimer::singleShot(requestIntervalTimer_+5, this, SLOT(disableMCAMCS()));
+		}
+		e->accept();
+		return true;
+	}
+	else if(e->type() == (QEvent::Type)AmptekEventDefinitions::ConfigurationRequestEvent){
+		if(((AmptekConfigurationRequestEvent*)e)->allParametersRequest_)
+			requestAllTextConfigurationParameters();
+		e->accept();
+		return true;
+	}
+	else if(e->type() == (QEvent::Type)AmptekEventDefinitions::DwellRequestEvent){
+		if(((AmptekDwellRequestEvent*)e)->dwellMode_)
+//			enableMCAMCS();
+			requestDataPacket(AmptekCommandManagerSGM::RequestEnableMCAMCS);
+		e->accept();
+		return true;
+	}
+	else if(e->type() == (QEvent::Type)AmptekEventDefinitions::ConfigurationSetEvent){
+		QStringList configurationCommands = ((AmptekConfigurationSetEvent*)e)->configurationCommands_;
+		textConfigurationSet(configurationCommands.at(0));
+		e->accept();
+		return true;
+	}
+	return QObject::event(e);
+}
+
+int AmptekSDD123Server::forwardDatagram(const QByteArray &datagram){
+	AmptekSDD123Packet forwardingPacket(packetIDCounter_++, datagram, true);
+	if(forwardingPacket.isValid()){
+		sendRequestDatagram(forwardingPacket);
+		return forwardingPacket.packetID();
+	}
+	return -1;
+}
+
+void AmptekSDD123Server::requestDataPacket(AmptekCommandManagerSGM::AmptekCommandDefSGM command, const QString &dataStringHex, bool sendFakeRequest, bool sendSyncRequest, int overrideTimeout)
+{
+	AmptekSDD123Packet requestStatusPacket(packetIDCounter_++, AmptekCommandManagerSGM::sddCommands()->amptekCommand(command).hex(), dataStringHex);
```

Comment:
This is really a minor style thing, but having an increment in the parameter passing of a function tends to make me edgy. When maintaining I always find some unnecessary thought needs to go into whether the value passed will be the value before increment, or after.
In this case I think:
```C++
AmptekSDD123Packet requestStatusPacket(packetIDCounter_, AmptekCommandManagerSGM::sddCommands()->amptekCommand(command).hex(), dataStringHex);
packedIDCounter_++;
```
...would be a little more obvious as to the intention of the programmer.